### PR TITLE
#635 Describe how to back up and restore projects for v0.6.x and 0.7.x

### LINKED ISSUE #635: How to backup and restore a Keptn installation
We need documentation for 0.6.x and 0.7.x on how to backup and restore Keptn data.

This is especially important for the following two Kubernetes deployments:

* mongodb (contains all cloud-events; used by mongodb-datastore)
* configuration-service (contains all config repos)

Please note: Between 0.6.x and 0.7.x we have moved mongodb from the `keptn-datastore` namespace into the `keptn` namespace (and with 0.7.1 the namespace name can be arbitrary)

**Use case**: As a user, I want to create a backup of the configuration and cloud-events that I use with Keptn. If I ever accidentally delete the Keptn namespace, I want to be able to reinstall Keptn and restore data.

**Why this came up**: What if the upgrade from 0.6.2 to 0.7 fails? We should provide the user with an option to revert to the previously working installation (e.g., 0.6.2).

# Required research

Properly backing up mongodb is non-trivial and needs to be researched. We want to avoid backing up an inconsistent state.

# What it feels like for the user

1. Create a new cluster
1. Install Keptn 0.6.x or 0.7.x
1. Create projects, onboard services, send cloud-events (e.g., new-artifact, start-evaluation, ...)
1. Take a look at Bridge: Note down projects, services and cloud-events
1. Create a backup according to our documentation
1. Uninstall Keptn (and make sure to delete everything from the `keptn` and/or `keptn-datastore` namespace)
1. Install Keptn 0.6.x or 0.7.x (same version as before)
1. Restore data from a backup according to our documentation
1. Access Bridge: All projects, services and cloud-events that have been visible before are visible again

# Out of scope

- Backing up helm related data
- Backing up istio related data
- Backing up Kubernetes namespaces, secrets, etc...
- Restoring data of an incompable Keptn version

# Definition of Done

- [ ] The backup documentation does not require any third party tools to be installed locally, except for `kubectl` and `bash`
- [ ] The backup documentation must be such that data is backed up to a local drive, and restored from a local drive (read: the backup should not be stored in the cluster).
- [ ] Documentation is provided for Keptn 0.6.x on how to backup and restore mongodb and configuration-service
- [ ] Documentation is provided for Keptn 0.7.x on how to backup and restore mongodb and configuration-service
- [ ] Upgrade Guide from 0.6.x to 0.7.x contains a link to the backup documentation (e.g., please create a local backup following this guide before proceeding with the upgrade)

### PR DESCRIPTION
**Pull request by bacherfl**
Closes #635

### CODE REVIEW

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+```console
+mkdir mongodb-backup
+chmod o+w mongodb-backup
+```
+
+1. Copy the data from the MongoDB to your local directory:
+
+```console
+MONGODB_USER=$(kubectl get secret mongodb-credentials -n keptn -ojsonpath={.data.user} | base64 -d)
```
Comment (agrimmer) [inline content/docs/0.6.0/installation/backup_and_restore/index.md:50]: The mongodb_user and password is actually not needed here. Right?

**Review by agrimmer — COMMENTED** [review]
Great!

File: content/docs/0.6.0/installation/backup_and_restore/index.md (hunk 1 above)
Comment (bacherfl) [reply content/docs/0.6.0/installation/backup_and_restore/index.md]: correct, in this case it's not needed. I have removed it

**Review by agrimmer — APPROVED** [review]
lgtm

**Comment by christian-kreuzberger-dtx** [comment]
The headings are weird (two different headings/title):
Screenshot from 2020-09-10 16-23-44
I think you can remove the additional heading "Backup Keptn"

**Comment by johannes-b** [comment]
Can you please add it to the "develop" folder for v.0.8.x too.

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+```console
+mkdir mongodb-backup
+chmod o+w mongodb-backup
+```
+
+1. Copy the data from the MongoDB to your local directory:
+
+```console
+kubectl exec svc/mongodb -n keptn -- mongodump --uri="mongodb://user:password@localhost:27017/keptn" --out=./dump
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.6.0/installation/backup_and_restore/index.md:48]: It should be `-n keptn-datastore` here, right?

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+mkdir mongodb-backup
+chmod o+w mongodb-backup
+```
+
+1. Copy the data from the MongoDB to your local directory:
+
+```console
+kubectl exec svc/mongodb -n keptn-datastore -- mongodump --uri="mongodb://user:password@localhost:27017/keptn" --out=./dump
+
+MONGODB_POD=$(kubectl get pods -n keptn-datastore -lrun=mongodb -ojsonpath='{.items[0].metadata.name}')
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.6.0/installation/backup_and_restore/index.md]: For me the label is `name=mongodb`, not `run=mongodb`.
So the command does not work. Instead, I used this:
```console
MONGODB_POD=$(kubectl get pods -n keptn-datastore -lname=mongodb -ojsonpath='{.items[0].metadata.name}')
```

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+chmod o+w mongodb-backup
+```
+
+1. Copy the data from the MongoDB to your local directory:
+
+```console
+kubectl exec svc/mongodb -n keptn-datastore -- mongodump --uri="mongodb://user:password@localhost:27017/keptn" --out=./dump
+
+MONGODB_POD=$(kubectl get pods -n keptn-datastore -lname=mongodb -ojsonpath='{.items[0].metadata.name}')
+kubectl cp keptn/$MONGODB_POD:dump ./mongodb-backup/ -c mongodb
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.6.0/installation/backup_and_restore/index.md:51]: The cp command fails for me:
```
Error from server (NotFound): pods "mongodb-678bcdc68-svmct" not found
```

Comment (bacherfl) [reply content/docs/0.6.0/installation/backup_and_restore/index.md]: has just been fixed

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+```
+
+## Backup Git credentials
+
+For projects where you have configured a Git upstream repository, we advise to backup the Kubernetes secrets needed to access those.
+To do so, please execute the following command **for each project** for which you have configured an upstream repo:
+
+```console
+PROJECT=<project-name>
+kubectl get secret -n keptn git-credentials-$PROJECT_NAME -oyaml > $PROJECT-credentials.yaml
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.6.0/installation/backup_and_restore/index.md:61]: Mismatch: Variable is called `PROJECT`, but you are accessing `PROJECT_NAME` here

File: content/docs/0.7.x/operate/backup_and_restore/index.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+kubectl cp keptn/$MONGODB_POD:dump ./mongodb-backup/ -c mongodb
+```
+
+## Backup Git credentials
+
+For projects where you have configured a Git upstream repository, we advise to backup the Kubernetes secrets needed to access those.
+To do so, please execute the following command for each project for which you have configured an upstream repo:
+
+```console
+kubectl get secret -n keptn git-credentials-$PROJECT_NAME -oyaml > $PROJECT-credentials.yaml
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.7.x/operate/backup_and_restore/index.md:62]: Mismatch: Variable is called `PROJECT`, but you are accessing `PROJECT_NAME` here

Also: `PROJECT` should be set here, same as in the 0.6.x docs:
```
PROJECT=<project-name>
```

File: content/docs/0.8.x/operate/backup_and_restore/index.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+kubectl cp keptn/$MONGODB_POD:dump ./mongodb-backup/ -c mongodb
+```
+
+## Backup Git credentials
+
+For projects where you have configured a Git upstream repository, we advise to backup the Kubernetes secrets needed to access those.
+To do so, please execute the following command for each project for which you have configured an upstream repo:
+
+```console
+kubectl get secret -n keptn git-credentials-$PROJECT_NAME -oyaml > $PROJECT-credentials.yaml
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.8.x/operate/backup_and_restore/index.md:62]: Mismatch: Variable is called `PROJECT`, but you are accessing `PROJECT_NAME` here

Also: `PROJECT` should be set here, same as in the 0.6.x docs:
```
PROJECT=<project-name>
```

File: content/docs/0.6.0/installation/backup_and_restore/index.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+```console
+mkdir keptn-backup
+cd keptn-backup
+mkdir config-svc-backup
+CONFIG_SERVICE_POD=$(kubectl get pods -n keptn -lrun=configuration-service -ojsonpath='{.items[0].metadata.name}')
+kubectl cp keptn/$CONFIG_SERVICE_POD:/data ./config-svc-backup/ -c configuration-service
+```
+
+1. Verify that the data has been copied correctly to your local machine by checking the content of the directory you just created.
```
Comment (christian-kreuzberger-dtx) [inline content/docs/0.6.0/installation/backup_and_restore/index.md:25]: The enumeration with `1.` does not seem to work.

I think there should not be a newline between the two lists. Anyway, you can use `2.` here too.

**Review by christian-kreuzberger-dtx — CHANGES_REQUESTED** [review]
From a technical point of view, I tested this on 0.6.2 and 0.7.1 and it works now, good job :)

From a formatting point of view, some enumerations or bullet points seem to be weird or wrong, see this screenshot:

Screenshot from 2020-09-11 14-38-23


I think that's the last thing to fix, then we can get this merged!

**Review by christian-kreuzberger-dtx — APPROVED** [review]
lgtm

**Pull request merged**